src/coreclr/src/vm/threads.h

```diff
@@ -1958,7 +1958,14 @@ class Thread
         //   ** WARNING ** WARNING ** WARNING ** WARNING ** WARNING ** WARNING **  |
         // ------------------------------------------------------------------------
 
+
+#if defined(TARGET_AMD64) || defined(TARGET_X86)
+        // on the Intel architecture we have strong memory ordering guarantees
         m_fPreemptiveGCDisabled.StoreWithoutBarrier(1);
+#else
+        // weaker memory models need an interlocked operation to ensure consistency
+        FastInterlockOr(&m_fPreemptiveGCDisabled, 1);
```

Comment:
We should not be using any barrier here even for weak architectures. It would have significant performance consequences.  @jkotas mentioned just recently that it is not unusual to see workloads that flip this flag million or more times per second. Reading that flag happens on many orders slower rate, basically just during GC suspension.
So our synchronization model is based on writing to the m_fPreemptiveGCDisabled without any barrier and executing process wide memory barrier (FlushProcessWriteBuffers) at the read time.

Reply:
right we don't need to do this. if a barrier is actually needed (which I'm not sure that it is...still thinking), we can add one on the GC side only when a suspension is required.

---
In reply to: [472863596](https://github.com/dotnet/runtime/pull/40931#discussion_r472863596) [](ancestors = 472863596)